# What are Operators?

Operators in Python are special symbols that perform operations on variables and values. They are used to performarithmetic operations, logical operations, and comparisons.

1. Arithmetic Operators 
2. Identity Operators
3. Assignment Operators 
4. Membership Operators
5. Comparison Operators 
6. Bitwise Operators
7. Logical Operators

In [ ]:
# 1. Arithmetic Operators
a = 10
b = 3
print(a + b)    # 13 (Addition)
print(a - b)    # 7  (Subtraction)
print(a * b)    # 30 (Multiplication)
print(a / b)    # 3.3333... (Division)
print(a % b)    # 1  (Modulus)
print(a ** 2)   # 100 (Exponentiation)
print(a // b)   # 3  (Floor division)

# 2. Identity Operators
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x
print(x is z)     # True 
print(x is y)     # False 
print(x is not y) # True

# 3. Assignment Operators
n = 5        # n = 5
n += 2       # n = n + 2 
n -= 1       # n = n - 1 
n *= 3       # n = n * 3 
n /= 2       # n = n / 2 

# 4. Membership Operators
fruits = ["apple", "banana", "cherry"]
print("apple" in fruits)      # True
print("orange" not in fruits)  # True

# 5. Comparison Operators
i = 10
j = 20
print(i == j)  # False 
print(i != j)  # True 
print(i > j)   # False 
print(i < j)   # True 
print(i >= 10) # True 

# 6. Bitwise Operators
# 5 = 0101 (Binary), 3 = 0011 (Binary)
print(5 & 3)  # 1  (AND: 0001)
print(5 | 3)  # 7  (OR: 0111)
print(5 ^ 3)  # 6  (XOR: 0110)
print(~5)     # -6 (NOT)
print(5 << 1) # 10 (Left Shift)

# 7. Logical Operators
age = 25
print(age > 18 and age < 30) # True 
print(age > 20 or age < 15)  # True 
print(not(age == 25))        # False 

- The arithmetic operators return the type of result dependson the type of operands, as below.
    - If either operand is a complex number, the result is converted to complex;
    - If either operand is a floating point number, the result is converted to floating point;

# 1. Arithmetic Operators

In [1]:
a = 10
b = 3
print(a + b)    # 13 (Addition)
print(a - b)    # 7  (Subtraction)
print(a * b)    # 30 (Multiplication)
print(a / b)    # 3.3333... (Division)
print(a % b)    # 1  (Modulus)
print(a ** 2)   # 100 (Exponentiation)
print(a // b)   # 3  (Floor division)


13
7
30
3.3333333333333335
1
100
3


# 2. Assignment Operators

- The assignment operators are used to assign values to variables.
- According to PEP 8 — Python’s official style guide: "Surround binary operators with a single space on both sides."

In [2]:
# assignment
x = 5
print(x)

# add and assign
x = 5
x += 7
print(x)

# subtract and assign
x = 5
x -= 7
print(x)

# multiply and assign
x = 5
x *= 7
print(x)

# divide and assign
x = 5
x /= 7
print(x)

# modulo and assign
x = 5
x %= 7
print(x)

# exponentiation and assign
x = 5
x **= 7
print(x)

# floor division and assign
x = 5
x //= 7
print(x)

5
12
-2
35
0.7142857142857143
5
78125
0


# 3. Comparison Operators

- Comparison operators are used to compare two values or variables. The result of a comparison is always a Boolean value: either True or False.

In [1]:
a = 10
b = 3

print(a == b)  # Output: False
print(a != b)  # Output: True
print(a > b)   # Output: True
print(a <= 5)  # Output: False

False
True
True
False


# 4. Logical Operators

## 4.1 AND Operator

- The and operator returns True only if both conditions are true

In [ ]:
x = 10
y = 20

result = (x < 15) and (y > 5)
print(result) 

True


## 4.2 OR Operator


- The or operator returns True if at least one of the conditions is true. It only returns False if both conditions are false.

In [ ]:
x = 10
y = 20

result_1 = (x < 15) or (y < 25)
result_2 = (x < 8) or (y < 25)
result_3 = (x < 8) or (y < 15)
print(result_1)
print(result_2)
print(result_3)

True
True
False


## 4.3 NOT Operator

- Reverse the result, returns False if the result is true.

In [9]:
x = 5
print(not x < 10)

False


# 5. Identity Operator

## 5.1 is Operator

- we are using Indentity Operators for comparing the memery locations of variables 

- Returns True if both variables refer to the same object
- Key Difference: == vs is
    - **==** (Comparison): Checks if the values are the same.
    - **is** (Identity): Checks if the IDs (memory addresses) are the same.

In [ ]:
x = 10
y = 20
print(x is y)  

False


In [12]:
x = 10
y = 10
print(x is y)
print(id(x))
print(id(y))

True
140726819247512
140726819247512


**The reason the result is True in your example is that both variables x and y are pointing to the exact same memory location**
- In Python, the is operator is an Identity Operator. Unlike the == operator (which checks if values are equal), the is operator checks if two variables share the same identity (memory address).

- When you assign a small integer (like 10) to multiple variables, Python does not create a new object in memory for each one. Instead:
    - It creates the object 10 once.
    - It points both x and y to that same pre-existing memory address to save space.

```
x = 10
y = 10
print(x is y)
print(id(x))
print(id(y))
```
- output : 
    - True
    - 140726819247512
    - 140726819247512

- **Small Integers (-5 to 256): Are interned (reused), so is usually returns True.**

## 5.2 is not

- The is not operator returns True if two variables point to different objects in memory, even if their values are identical.Key ConceptPurpose: It checks for "Non-Identity" (different memory addresses/IDs).Result: It returns True if id(x) != id(y).

In [13]:
x = 500
y = 500
print(x is not y) # Output: True (Usually stored at different memory addresses)

True


# 6. Membership Operator  

 - Membership Operators are used to test if a specific value or variable is found within a sequence (such as a string, list, tuple, set, or dictionary).

## 6.1 in Operator

- The in operator returns True if the specified value is present in the sequence.

In [ ]:
sports = ["tennis", "cricket", "football"]
print("cricket" in sports) 

True


In [15]:
this_dict = {
    "name": "Sugar",
    "weight": "1kg",
    "price": 130.35
}

print("price" in this_dict)   # Output: True (Checks keys)
print(130.35 in this_dict)    # Output: False (Does not check values)

True
False


## 6.2 not in Operator

- The not in operator returns True if the specified value is not present in the sequence.

In [ ]:
sports = ["tennis", "cricket", "football"]
print("rugby" not in sports) 


True


# 7. Bitwise Operator 

- Based on your lecture materials, Bitwise Operators are used to perform operations on binary numbers (0s and 1s). While we usually write numbers in decimal (base 10), Python converts them to binary to perform these specific operations

1. Bitwise AND => &
2. Bitwise OR => |
3. Bitwise XOR => ^
4. Bitwise NOR => `

## 7.1 Bitwise AND => &

- Computers see numbers in binary (0s and 1s). When you use &, Python performs a comparison for every single bit.
    - 1 & 1 = 1
    - 1 & 0 = 0
    - 0 & 1 = 0
    - 0 & 0 = 0

In [ ]:
x = 5
y = 3
result = x & y
print(result) 

1


## 7.2 Bitwise OR => |

- When you use the | operator, Python converts the decimal numbers into binary and follows this logic for every bit position:
    - 1 | 1 = 1
    - 1 | 0 = 1
    - 0 | 1 = 1
    - 0 | 0 = 0

In [1]:
x = 5
y = 3
result = x | y
print(result) # Output: 7

7


## 7.3 Bitwise XOR => ^

- The XOR (Exclusive OR) operator sets each bit to 1 only if one of the two bits is 1. If both bits are 1 or both are 0, it returns 0.

    - 1 ^ 1 = 0
    - 1 ^ 0 = 1
    - 0 ^ 1 = 1
    - 0 ^ 0 = 0

In [3]:
a = 10
b = 7
result = a ^ b

print(result) 
# Output: 13

13


## 7.4 Bitwise NOT => `

- The NOT operator inverts all the bits of a number. It turns 1s into 0s and 0s into 1s.
    - ~1 = 0
    - ~0 = 1

- **In Python, bitwise NOT of a number $x$ returns $-(x + 1)$. This happens because computers use a system called Two's Complement to store signed integers**

In [2]:
x = 10
print(~x) # Output: -11

-11
